python/cuml/sparse/linalg/blopex_lobpcg.py

```diff
@@ -0,0 +1,647 @@
+        VBV = cp.transpose(cholesky(VBV))
+        VBV = inv(VBV)
+        blockVectorV = cp.matmul(blockVectorV, VBV)
+        if B is not None:
+            blockVectorBV = cp.matmul(blockVectorBV, VBV)
+        else:
+            blockVectorBV = None
+    except Exception as e:
+        # raise ValueError('Cholesky has failed')
+        print("{} occured".format(e))
+        blockVectorV = None
+        blockVectorBV = None
+        VBV = None
+
+    if retInvR:
+        return blockVectorV, blockVectorBV, VBV, normalization
+    else:
+        return blockVectorV, blockVectorBV
+
+
+def _get_indx(_lambda, num, largest):
+    """Get `num` indices into `_lambda` depending on `largest` option."""
+    ii = cp.argsort(_lambda)
+    if largest:
+        ii = ii[:-num - 1:-1]
+    else:
+        ii = ii[:num]
+    return ii
+
+
+def _genEigh(A, B):
+    """
+    Helper function for converting a generalized eigenvalue problem
+    AX = lambdaBX to standard using cholesky.
+    This is because cupy does not have a functional api to solve
+    generalized eigenvalue problem.
+    Factorizing B = R^TR. Let F = (R^T)^-1 A R^-1
+    Equivalent Standard form: Fy = lambda(y), where our required
+    eigvec x = R^-1 y
+    """
+    # we transpose lower triangular matrix to get upper triangular matrix
+    R = cp.transpose(cholesky(B))
+    RTi = inv(cp.transpose(R))
+    Ri = inv(R)
+    F = cp.matmul(RTi, cp.matmul(A, Ri))
+    vals, vecs = eigh(F)
+    eigVec = cp.matmul(Ri, vecs)
+
+    return vals, eigVec
+
+
+def _bmat(list_obj):
+    """
+    Helper function to create a block matrix in cupy from a list
+    of smaller 2D matrices using iterated vertical and horizontal stacking
+    """
+
+    arr_rows = cp.array([])
+    for row in list_obj:  # row & list_obj are a list of cupy arrays
+        arr_cols = cp.array([])
+        for col in row:  # col is a cupy ndarray
+            if col.ndim == 0:
+                col = cp.array([col])
+            if(arr_cols.size == 0):
+                arr_cols = col
+                continue
+
+            arr_cols = cp.hstack((arr_cols, col))
+        if(arr_rows.size == 0):
+            arr_rows = arr_cols
+            continue
+        arr_rows = cp.vstack((arr_rows, arr_cols))
+
+    return arr_rows
+
+
+def _handle_gramA_gramB_verbosity(gramA, gramB, verbosityLevel):
+    if verbosityLevel > 0:
+        _report_nonhermitian(gramA, 'gramA')
+        _report_nonhermitian(gramB, 'gramB')
+
+
+def lobpcg(A,
+           X,
+           B=None,
+           M=None,
+           Y=None,
+           tol=None,
+           maxiter=None,
+           largest=True,
+           verbosityLevel=0,
+           retLambdaHistory=False,
+           retResidualNormsHistory=False):
+    """Locally Optimal Block Preconditioned Conjugate Gradient Method (LOBPCG)
+    LOBPCG is a preconditioned eigensolver for large symmetric positive
+    definite (SPD) generalized eigenproblems.
+    Parameters
+    ----------
+    A : {cupy.ndarray}
```

Comment:
We are able to support cupyx.scipy.sparse arrays here too, right?

Reply:
Technically yes, but `linearOperator` abstracts this out and enables matmuls in scipy. The same cannot be done in cupy (as of now) so one has to densify, then do a matmul. Once 'linearOperator` support is added (which is next on the agenda), `cupyx.scipy.sparse` arrays will get automatically supported.